### PR TITLE
Vault Spawning Point System

### LINKED ISSUE #27385: Feature Request: Change how vaults are spawned
**Reasoning**
Right now vaults are just spawned randomly from a list. On startup the server decides how many will spawn then just picks that many. Problem is that not all vaults are made equally. On paper this just adds RNG and makes vault hunting exciting but not necessarily round changing every time. In practice it makes it so adding several small, less interesting vaults actively makes vault hunting less fun because your cute little garden or space bakery that's really very charming has overwritten spessmart. 

**Suggestion**
A point system, make small or less rewarding vaults cost few points, make large vaults like spessmart or the zombie one cost a lot of points. We don't have enough vaults to make retro-actively adding a point value much effort at all. This way if 10 small flavor vaults are added it won't make anyone's round boring after the first time they see them. This means space can be more populated without just raw powercreep. 

I'd do this myself but I'm not confident enough tinkering with things related to the server itself in this regard. If it gets added I'll likely pump out some small vaults though. 

[vault] [system]

### PR DESCRIPTION
**Pull request by GreatBigFailure**
This is something I've been wanting for years now, and the vault contest was a great excuse to do it myself.

### What this changes

Vaults currently spawn completely randomly with equal weight, a number between 5 and 15 is picked, and that's how many vaults you get. 

What this does is add a point value to every vault, which is then subtracted from a total (RNG based) point value when spawning. Meaning **more small scenery-based vaults can spawn, while not being in (as much) direct competition with larger round altering vaults**, so both scenery enjoyers and loot piñata hunters benefit.  

### The math

Including this for transparency, this will increase the total number of vaults that spawn by a noticable, but not too significant number. 
Old: **Number between 5 - 15 = amount of vaults**
New: **Number between 5 - 15, multiplied by 3 = amount of vault spawn points.** 
This is based on a **1 - 5 point scale. So the multiplier of 3 is meant to be an "average" vault cost, effectively equal to what we do now.

### Big List of Vault Point Costs
Balance was based around danger, time investment, rewards, and in some cases chance of finding/engaging with it at all. Also personal opinion.

**1 Point**
- Ice truck crash
- Tommy boy asteroid
- Space gym
- Space Pond
- IOU vault
- Hivebot crash
- AI satellite
- Ice comet
- Beach party
- Zathura

**2 Points**
- Asteroid Temple
- Mechanic Clubhouse
- RUST
- Dance Dance Revolution
- Biodome
- Asteroids
- Prison Ship
- Taxi Engineering
- Asteroid Field
- Pod Station
- Mini Station
- Fast Food Joint

**3 Points**
- Broke UFO
- Lightship
- Hivebot Factory
- Clown Base
- Listening Satellite
- Drone Ship
- AME Lab
- Zoo Truck
- Skeleton Den
- Spy Satellite
- Iron Chef
- Assistant's Lair
- Clown Roid
- Goon Sat
- Laundromat
- Keycard Vault
- Sokoban
 
**4 Points**
- Gingerbread House
- Ejected Engine
- Meteorlogical Station
- Research Facility
- Syndie Cargo

**5 Points**
- Mothership Lab
- Black Site Prism
- Old Armory
- Spessmart

I'm not 100% set on the balance of these, and I might be overlooking something so please let me know if I forgot some 2 point vault has something that should make it 4 points or something.

Finally, this also fixed an oversight that prevented the Laundromat vault from spawning.

Closes #27385 

:cl:
 * rscadd: Introduces a point cost system for vault spawning, so more small vaults can spawn without replacing larger ones.
 * bugfix: The Laundromat vault can now spawn properly
 * bugfix: The AME vault will no longer break due to rotating

### CODE REVIEW

**Comment by GreatBigFailure** [comment]
Now that I've listed them all out I think I should change the 3.5 multiplier to just 3, there's a lot more 3 or less point than 4 and 5. This might change with the contest but then it can be bumped up if needed. 

I'll wait for people's comments to do that so I can tweak multiple things at once.

**Comment by gurfan** [comment]
The average cost here is about 2.5. I think you should lower the multiplier to 3 unless you want a lot more vaults to be spawning

**Comment by gurfan** [comment]
Black site prism, Syndie cargo, and whichever vault has the syndicate beacon in it (old armory? spy satellite?) should all be 5 points.

And ejected engine can probably be bumped down to 3 imo.

**Comment by GreatBigFailure** [comment]
> Black site prism, Syndie cargo, and whichever vault has the syndicate beacon in it (old armory? spy satellite?) should all be 5 points.

My mindset with them being at 4 is that most people just end up dead when they try them, so a lot of people skip over them. You're probably right though, I'll double check the contents and bump them up, at least the larger of the two syndie vaults.

> And ejected engine can probably be bumped down to 3 imo.

Flawless bluespace crystal and 2 free large artifacts, but it isn't 10,000,000 guns, so you're probably right there as well.

**Comment by GreatBigFailure** [comment]
Okay, I changed the multiplier from 3.5 to 3. I also bumped black prism site up to 5 points, I underestimated exactly how much stuff was in there. I bumped the old armory (larger syndicate ship) to 5 as well but kept the smaller one at 4. They both have a fair number of guns but being above the average point cost is enough for the smaller one. 

Undecided on the ejected engine, and by that I mean I forgot to change it so I'll wait for more feedback.

**Comment by Eneocho** [comment]
does the keycard vault spawn properly with this system?

**Comment by GreatBigFailure** [comment]
> does the keycard vault spawn properly with this system?

Yep

**Comment by Kurfursten** [comment]
Is there a chance that this could cause an infinite loop? Say by random chance it picks all 10 one-cost vaults and a couple 2-cost, but we run out of vault spawn space. Would it keep trying to spawn a vault because it hasn't fulfilled its point quota yet and thus hang the server?

** Potentially even more danger if we get more flavor vaults from the contest and some are physically large

**Comment by GreatBigFailure** [comment]
@Kurfursten I'll double check the way vault spawns/placements are handled. I remember there being fail safes in place for similar situations, but better safe than sorry.

Edit: Yes there's a safety, it'll be fiiiiine

**Comment by GreatBigFailure** [comment]
Took the chance to fix a bug with the AME vault (Just turned off rotation for it), closes #35763 
Also reduced ejected engine's point cost to 3 as per request

File: code/modules/randomMaps/vault_definitions.dm [hunk 1]
```diff
@@ -125,26 +138,32 @@ var/list/existing_vaults = list()
...
+	spawn_cost = 1
 
 /datum/map_element/vault/biodome
 	file_path = "maps/randomvaults/biodome.dmm"
+	spawn_cost = 2
 
-/datum/map_element/vault/iou_vault
-	file_path = "maps/randomvaults/iou_fort.dmm"
+///datum/map_element/vault/iou_vault
+//	file_path = "maps/randomvaults/iou_fort.dmm"	//Why was this in here twice?
```
Comment (Kurfursten) [inline code/modules/randomMaps/vault_definitions.dm]: just delete it

**Pull request merged**